### PR TITLE
NextJS error handling 

### PR DESCRIPTION
**Pull request by crysmags**
### What does this PR do?
This PR is to add error handling for NextJS - this will add the error stack to the trace as well as bubble up the error to the root span

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/datadog-instrumentations/src/next.js [hunk 1]
```diff
@@ -154,6 +155,11 @@ function finish (ctx, result, err) {
     errorChannel.publish(ctx)
   }
 
+  const maybeNextRequest = nodeNextRequestsToNextRequests.get(ctx.req)
+  if (maybeNextRequest) {
+    ctx = { nextRequest: maybeNextRequest, ...ctx }
```
Comment (rochdev) [inline packages/datadog-instrumentations/src/next.js:160]: The `ctx` object must remain the same across calls and shouldn't be replaced here.

File: packages/datadog-plugin-next/src/index.js [hunk 2]
```diff
@@ -43,16 +43,17 @@ class NextPlugin extends ServerPlugin {
...
     const store = storage.getStore()
 
     if (!store) return
 
     const span = store.span
     const error = span.context()._tags['error']
 
-    if (!this.config.validateStatus(res.statusCode) && !error) {
-      span.setTag('error', true)
+    iif (!this.config.validateStatus(res.statusCode) && !error) {
```
Comment (rochdev) [inline packages/datadog-plugin-next/src/index.js:54]:
```suggestion
    if (!this.config.validateStatus(res.statusCode) && !error) {
```

File: packages/datadog-plugin-next/test/app/api/appDir/error/route.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { NextResponse } from 'next/server'
+
+export async function GET (req) {
```
Comment (rochdev) [inline packages/datadog-plugin-next/test/app/api/appDir/error/route.js:3]: Where is this run?

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-instrumentations/src/next.js (hunk 1 above)
Comment (sabrenner) [reply packages/datadog-instrumentations/src/next.js:160]: Would changing it to `ctx.nextRequest = maybeNextRequest` be ok?

Comment (rochdev) [reply packages/datadog-instrumentations/src/next.js:160]: Yes, that's basically the expected use of `ctx`. The idea is that having the same object everywhere allows events to share data.

File: packages/datadog-plugin-next/test/app/api/appDir/error/route.js (hunk 3 above)
Comment (rochdev) [reply packages/datadog-plugin-next/test/app/api/appDir/error/route.js:3]: Let's convert this to a draft if it's not ready for review.

**Review by Qard — APPROVED** [review]

**Review by rochdev — APPROVED** [review]

**Pull request merged**